src/librustc_metadata/decoder.rs

```diff
@@ -1099,20 +1094,28 @@ fn doc_predicates<'a, 'tcx>(base_doc: rbml::Doc,
 {
     let mut dcx = base_doc.get(tag).decoder();
     dcx.cdata = Some(cdata);
+    dcx.tcx = Some(tcx);
 
     ty::GenericPredicates {
         parent: dcx.decode(),
-        predicates: dcx.seq().map(|offset| {
-            let predicate_pos = cdata.xref_index.lookup(
-                cdata.data(), offset).unwrap() as usize;
-            let mut dcx = rbml::Doc {
-                data: cdata.data(),
-                start: predicate_pos,
-                end: cdata.data().len(),
-            }.decoder();
-            dcx.tcx = Some(tcx);
-            dcx.cdata = Some(cdata);
-            dcx.decode()
+        predicates: (0..dcx.decode::<usize>()).map(|_| {
+            // Handle shorthands first, if we have an usize > 0x80.
+            if dcx.opaque.data[dcx.opaque.position()] & 0x80 != 0 {
```

Comment:
Nit: it surprises me to see this open-coded here; I expect some sort of helper to handle this generically